### PR TITLE
Refactor `MSAppDelegateForwarder` to `MSDelegateForwarder` to allow inheritance

### PR DESCRIPTION
**Pull request by dhei**
* ~[ ] Has `CHANGELOG.md` been updated?~
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Refactor work without functional change to prepare for implementation of UNUserNotificationCenterDelegate delegate forwarder.
Abstract away delegate forwarder into `MSDelegateForwarder` to allow inheritance of `MSAppDelegateForwarder`.

### CODE REVIEW

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSCustomApplicationDelegate.h [hunk 1]
```diff
@@ -9,7 +10,7 @@ NS_ASSUME_NONNULL_BEGIN
  * registered through swizzling. Due to the early registration of swizzling on the original app delegate each custom delegate must sign up
  * for selectors to swizzle within the `load` method of a category over the @see MSAppDelegateForwarder class.
  */
-@protocol MSCustomApplicationDelegate <NSObject>
+@protocol MSCustomApplicationDelegate <MSCustomDelegate>
```
[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSCustomApplicationDelegate.h]: Yes, `MSCustomDelegate` is the base protocol now since we added an inheritance layer to the AppDelegate it makes sens to do the same on the protocol too. When we'll introduce `MSCustomUserNotificationCenterDelegate` it will also inherit from `MSCustomDelegate`.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+#pragma mark - Custom UIApplicationDelegate
+
+#if !TARGET_OS_OSX
+
+/*
+ * Those methods will never get called but their implementation will be used by swizzling. Those implementations will run within the
+ * delegate context. Meaning that `self` will point to the original app delegate and not this forwarder.
+ */
+- (BOOL)custom_application:(UIApplication *)application
```
[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: These are very specific  method because they are the place holder for the swizzled implementation, they'll never get actually called. I thought I'd differentiate them from classic methods but I am opened to discussion.

[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: Also the naming gets a special treatment it has to start with `kMSCustomSelectorPrefix` as a convention so swizzling can identify the custom version and the original version of the API. We can still update it if we need.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Hold the original setDelegate implementation.
+ */
+@property(nonatomic) IMP originalSetDelegateImp;
+
+/**
+ * Keep track of the original delegate's method implementations.
+ */
+@property(nonatomic, readonly) NSMutableDictionary<NSString *, NSValue *> *originalImplementations;
```
[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.h]: It's true, it's also probably faster, though I'll just add a TODO since that was already the case before. Also we manipulate the string directly [here ](https://github.com/Microsoft/AppCenter-SDK-Apple/pull/1191/files/2821e1e8e10b647653fb16b5fab01b11ba98b78b#diff-033759a6401f4841724c1b1c53ede8b9R112) but we can deal with it even with a value.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.m [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+/**
+ * Custom implementation of the setDelegate: method.
+ *
+ * @param delegate The delegate to be swizzled, its type here is @c id<NSObject> to be generic but your implementation will have to declare
+ * the exact type of the expected delegate (i.e.: @c MSApplicationDelegate).
+ *
+ * @discussion Beware, @c self in this method is not the current class but the swizzled class.
+ */
+- (void)custom_setDelegate:(__unused id<NSObject>)delegate {
```
[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.m]: Same reason than for `custom_application`.

File: AppCenter/AppCenter.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -409,6 +407,18 @@
...
 		35DFC2402170053100455589 /* MSTypedPropertyTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 35DFC21C2170044600455589 /* MSTypedPropertyTests.m */; };
+		38032087217E85A90089772A /* MSDelegateForwarder.h in Headers */ = {isa = PBXBuildFile; fileRef = 38032085217E85A90089772A /* MSDelegateForwarder.h */; };
+		38032088217E85A90089772A /* MSDelegateForwarder.h in Headers */ = {isa = PBXBuildFile; fileRef = 38032085217E85A90089772A /* MSDelegateForwarder.h */; };
+		38032089217E85A90089772A /* MSDelegateForwarder.h in Headers */ = {isa = PBXBuildFile; fileRef = 38032085217E85A90089772A /* MSDelegateForwarder.h */; };
+		3803208A217E85A90089772A /* MSDelegateForwarder.m in Sources */ = {isa = PBXBuildFile; fileRef = 38032086217E85A90089772A /* MSDelegateForwarder.m */; };
+		3803208B217E85A90089772A /* MSDelegateForwarder.m in Sources */ = {isa = PBXBuildFile; fileRef = 38032086217E85A90089772A /* MSDelegateForwarder.m */; };
+		3803208C217E85A90089772A /* MSDelegateForwarder.m in Sources */ = {isa = PBXBuildFile; fileRef = 38032086217E85A90089772A /* MSDelegateForwarder.m */; };
```
[the remark this answers is not in the document]
Comment (dhei) [reply AppCenter/AppCenter.xcodeproj/project.pbxproj]: @ElektrojungeMS They are not duplicates; they are 3 different UUIDs for 3 different targets: AppCenterIOS, AppCenterMacOS, and AppCenterTVOS.

3803208**7**217E85A90089772A
3803208**8**217E85A90089772A
3803208**9**217E85A90089772A

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (achocron) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: No worries then, just thought I'd ask

[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: I think we can safely rename now but not sure if we want to do it. The original reason was that the whole method was supposed to be added to the original delegate hence breaking the conventions to avoid conflicts then we moved to another safer technic where we store the implementations in cache. Either we probably won't rename in this PR.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+#pragma mark - Custom Application
+
+- (void)custom_setDelegate:(id<MSApplicationDelegate>)delegate {
+
+  // Swizzle only once.
+  static dispatch_once_t delegateSwizzleOnceToken;
+  dispatch_once(&delegateSwizzleOnceToken, ^{
+    // Swizzle the delegate object before it's actually set.
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: line break here

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.h [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#import "MSCustomApplicationDelegate.h"
+#import "MSDelegateForwarder.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * Enum used to represent all kind of executors running the completion handler.
+ */
+typedef NS_OPTIONS(NSUInteger, MSCompletionExecutor) {
```
Comment (jwallra) [inline AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.h]: Why are we using an enum but setting its values like flags?  Are these flags or values?  I can't tell.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.h [hunk 8]
```diff
@@ -0,0 +1,77 @@
+#import <Foundation/Foundation.h>
+@protocol MSCustomDelegate;
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.h]: line break here

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m (hunk 6 above)
Comment (jwallra) [inline AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: Newline before comment

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.m [hunk 9]
```diff
@@ -0,0 +1,259 @@
+#import "MSDelegateForwarder.h"
+#import "MSAppCenterInternal.h"
+#import "MSCustomDelegate.h"
+#import "MSDelegateForwarderPrivate.h"
+#import "MSLogger.h"
+
+#import <objc/runtime.h>
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.m]: `#import <...>` should come first. Is it caused by new formatter?

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.m [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    _originalImplementations = [NSMutableDictionary new];
+    _traceBuffer = [NSMutableArray new];
+    _enabled = YES;
+  }
+  return self;
+}
+
++ (instancetype)sharedInstance {
+
+  // This is an empty method and expect to be overridden in sub classes.
```
Comment (jwallra) [inline AppCenter/AppCenter/Internals/DelegateForwarder/MSDelegateForwarder.m]: "...and *is expected* to be..." would be correct here

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m [hunk 11]
```diff
@@ -765,8 +766,9 @@ - (void)storeDownloadedReleaseDetails:(nullable MSReleaseDetails *)details {
   [MS_USER_DEFAULTS setObject:groupId forKey:kMSDownloadedDistributionGroupIdKey];
   [MS_USER_DEFAULTS setObject:releaseId forKey:kMSDownloadedReleaseIdKey];
   [MS_USER_DEFAULTS setObject:releaseHashes forKey:kMSDownloadedReleaseHashKey];
-  MSLogDebug([MSDistribute logTag], @"Stored downloaded release hash(es) (%@) and id (%@) for later "
-                                    @"reporting.",
+  MSLogDebug([MSDistribute logTag],
+             @"Stored downloaded release hash(es) (%@) and id (%@) for later "
```
Comment (jaelim-ms) [inline AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: Formatter isn't smart enough to deal with this kind of format. It should be at least one less lines.

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m [hunk 12]
```diff
@@ -836,8 +839,9 @@ - (void)changeDistributionGroupIdAfterAppUpdateIfNeeded:(NSString *)currentInsta
   if ((storedDistributionGroupId == nil) || ([updatedReleaseDistributionGroupId isEqualToString:storedDistributionGroupId] == NO)) {
 
     // Set group ID from downloaded release details if an updated release was downloaded from another distribution group.
-    MSLogDebug([MSDistribute logTag], @"Stored group ID doesn't match the group ID of the updated "
-                                      @"release, updating group id: %@",
+    MSLogDebug([MSDistribute logTag],
+               @"Stored group ID doesn't match the group ID of the updated "
```
Comment (jaelim-ms) [inline AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: Same here. Please fix format for the line.

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m [hunk 13]
```diff
@@ -850,7 +854,6 @@ - (void)showConfirmationAlert:(MSReleaseDetails *)details {
 
   // Displaying alert dialog. Running on main thread.
   dispatch_async(dispatch_get_main_queue(), ^{
-
```
Comment (jaelim-ms) [inline AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: Revert this change.

File: AppCenterPush/AppCenterPush/MSPush.m [hunk 14]
```diff
@@ -220,9 +220,10 @@ - (void)registerForRemoteNotifications {
                               MSLogVerbose([MSPush logTag], @"Push notifications authorization was denied.");
                             }
                             if (error) {
-                              MSLogWarning([MSPush logTag], @"Push notifications authorization "
-                                                            @"request has been finished with "
-                                                            @"error: %@",
+                              MSLogWarning([MSPush logTag],
```
Comment (jaelim-ms) [inline AppCenterPush/AppCenterPush/MSPush.m]: Same here.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m (hunk 6 above)
Comment (dhei) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: clang format automatically removes the line break following the block syntax `^{` and there is no way to disable that behavior. 

I am fine to just conform with formatter here, but I am open to other suggestions.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.h (hunk 7 above)
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.h]: It's bitmask thanks to `NS_OPTIONS`, I can update the description.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m (hunk 6 above)
Comment (clpolet) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m]: Unfortunately the compiler doesn't like new lines at beginning of blocks but that's a separate issue.

File: AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.h (hunk 7 above)
Comment (jwallra) [reply AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.h]: I wrote this comment before I knew about the quirks of Common Schema, I think this is just a weird implementation that we'll have to deal with

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m (hunk 13 above)
Comment (clpolet) [reply AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: It'll pop up in another PR so it should be fine to have it here.
